### ISSUE TITLE
transforms: kornia -> torchvision?

### ISSUE THREAD

**Issue by adamjstewart** [torchgeo/torchgeo#3108]
### Summary

There have been many recent requests to switch from kornia to torchvision for our preprocessing/data augmentation transforms. This issue is designed to centralize all debate on this topic and collect feedback from developers and users.

### Rationale

This discussion started as a result of efforts to make all TorchGeo models (and their associated transforms) exportable to ONNX or .pt2 format. This feature is quite important to a subset of our users. There have also been some frustrations around the complexity of Kornia, especially for video data.

### Implementation

Kornia and torchvision are quite different. Kornia supports both sequential inputs with enum `data_keys` to denote the type of data as well as dicts with keys controlling input type. Torchvision uses TVTensor objects to label certain types and generally uses different dict keys than Kornia. We have standardized a lot of TorchGeo datasets and trainers to follow Kornia key names. Changing now would involve a lot of backwards-incompatible changes.

One way to minimize these backwards-incompatible changes would be to convert all dataset return values to TVTensors but keep all dict keys the same. This may make things confusing for torchvision users switching to TorchGeo (who have to learn new keys), but less confusing for existing TorchGeo users (who can upgrade without any breaking changes).

Model preprocessing transforms have already been ported in https://github.com/torchgeo/torchgeo/pull/2893. We would need a similar PR for data modules.

### Alternatives

An updateable list of pros/cons of possible alternatives. Hoping to keep these mostly objective.

### Kornia

* ✅ Backwards-compatibility across all TorchGeo releases if we don't change anything
* ✅ (subjective) good past experiences with upstreaming changes and adding features
* ✅ (subjective) sunk cost fallacy, we have already invested a lot in adding features we need
* ❌ Automatically changes dimensions from sample to batch, not always possible to prevent
* ❌ Some transforms require masks to have a fake channel dimension
* ❌ Not exportable, can't export to ONNX or .pt2 format
* ❌ Deep nested class hierarchy, harder to make major changes
* ❌ (subjective) VideoSequential is quite buggy, does not always behave as expected
* ❌ (subjective) harder to write custom transforms

### torchvision.transforms.v1

* ❌ Many transforms operate on PIL Images, which do not support SAR/MSI/HSI

### torchvision.transforms.v2

* ✅ Simple nn.Module subclass
* ✅ Exportable, no graph breaks
* ✅ (subjective) easier to write custom transforms
* ❌ (subjective) a bit harder to contribute new features/ideas, I've had more PRs rejected than accepted
* ❌ (subjective) "v2" makes this feel experimental, questions of stability

### Albumentations

From https://github.com/albumentations-team/albumentations#1-continue-using-albumentations-mit-license:

* ✅ Forever free for all uses including commercial
* ✅ No licensing fees or restrictions
* ❌ No bug fixes - Even critical bugs won't be addressed
* ❌ No new features - Missing out on performance improvements
* ❌ No support - Issues and questions go unanswered
* ❌ No compatibility updates - May break with new Python/PyTorch versions

The repo is archived and no longer maintained, which is likely a dealbreaker for us.

### AlbumentationsX

From https://github.com/albumentations-team/albumentations#2-upgrade-to-albumentationsx-dual-licensed

* ✅ Drop-in replacement - Same API, just pip install albumentationsx
* ✅ Active development - Regular updates and new features
* ✅ Bug fixes - Issues are actively addressed
* ✅ Performance improvements - Faster execution
* ✅ Community support - Active Discord and issue tracking
* ⚠️ Dual licensed:
  * ❌ AGPL-3.0: Free ONLY for projects licensed under AGPL-3.0 (not compatible with MIT, Apache, BSD, etc.)
  * ❌ Commercial License: Required for proprietary use AND permissive open-source projects

AGPL is not compatible with our MIT license, and we have no intentions to relicense TorchGeo. We don't have the money to afford a commercial license for all our users. This is definitely a dealbreaker for us.

### Additional information

I would like to keep all discussion related to this topic here on public GitHub as opposed to Slack. I propose performing a detailed pro/con analysis of all libraries, similar to how we chose to switch from R-tree to geopandas. I'm curious which library has the most transforms, whether any libraries are missing transforms we need, how performance (speed, memory) compares across libraries, how reproducible/stable these libraries are, etc. We need to make this decision before any future 1.0 release in order to ensure backwards-compatibility.

**Comment by isaaccorley** [comment]
Albumentations works on numpy arrays so no GPU support. It also doesn't support batch operations, so you would have to loop over a batch of samples.

**Comment by andrewaf1** [comment]
Torchvision transforms, to my knowledge, apply the same transform parameters to every item in a batch, while Kornia treats the items differently and applies different parameters by default. I don't know what effect this has on deep learning in general (seems to work fine for many people), but it does pose an issue for some SSL models like DinoV2 where you want to generate multiple views of the same scene by resize-cropping and augmenting one image into many (ideally using one large batch instead of multiple calls to the same transform).

**Comment by isaaccorley** [comment]
@andrewaf1 Great point! This is a major issue actually and win for kornia augmentations. For inference time transforms I still think that torchvision being exportable is a big deal. So likely need to consider just supporting kornia for training augmentations and torchvision for fixed inference transforms that aren't meant to build invariance during a training process.

I did also verify that this is still an issue in `torchvision.transforms.v2`

```python
import urllib.request
from io import BytesIO

import matplotlib.pyplot as plt
from PIL import Image
import torchvision.transforms.v2 as T
import torchvision.transforms.v2.functional as TF

transforms = T.Compose([
    T.Resize((224, 224)),
    T.RandomHorizontalFlip(p=0.5),
    T.RandomGrayscale(p=0.2),
])

url = "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS9Rp1dYzfE2UeJiEG9D-upcd5rtNLnMzvpYQ&s"
response = urllib.request.urlopen(url)
image = TF.pil_to_tensor(Image.open(BytesIO(response.read())))

# create a repeated batch of images
batch = image.repeat(4, 1, 1, 1)  # shape: (4, 3, H, W)
aug = transforms(batch)

# plot original and augmented images
fig, axs = plt.subplots(2, 4, figsize=(12, 6))
for i in range(4):
    axs[0, i].imshow(batch[i].permute(1, 2, 0))
    axs[0, i].axis('off')
    axs[0, i].set_title('Original')
    
    axs[1, i].imshow(aug[i].permute(1, 2, 0))
    axs[1, i].axis('off')
    axs[1, i].set_title('Augmented')
plt.tight_layout()
```

<img width="1171" height="594" alt="Image" src="https://github.com/user-attachments/assets/0995e199-2f85-4dde-8e3e-7a5501aebd04" />

**Comment by edgarriba** [comment]
For kornia missing features we are pretty open to land them if useful for the community @ducha-aiki @shijianjian

**Comment by ducha-aiki** [comment]
Sure we can help land some features. Just out of curiosity - why is onnx export needed? I am not sure if it supports the randomness, or is it about the transformations themselves?

**Comment by adamjstewart** [comment]
For ONNX export, we are specifically interested in exporting models with their preprocessing transforms (e.g., CenterCrop, Resize, Normalize). So all of the transforms without randomness.

**Comment by ducha-aiki** [comment]
This makes sense. I believe we can make them exportable. Do you have a specific export example to take a look at?

**Comment by adamjstewart** [comment]
@isaaccorley does, I'm guessing something like:
```python
import kornia.augmentation as K
import torch

aug = K.Normalize(0, 1)
inputs = torch.randn(1, 3, 224, 224)
torch.export.export(aug, inputs)
```

**Comment by shijianjian** [comment]
IIRC, there were ONNX PRs in those augmentations from your side previously.

Also, we have a fix for video sequential that had been merged several days ago. 

Probably you can try angain and list what works and not works for you.

**Comment by isaaccorley** [comment]
> Sure we can help land some features. Just out of curiosity - why is onnx export needed? I am not sure if it supports the randomness, or is it about the transformations themselves?

It wasn't specifically ONNX export, but torch export. Although ONNX support would also be nice because you could run these in webapps using the onnxruntime-web library.

**Comment by robmarkcole** [comment]
@isaaccorley keen to revive this thread. Is ONNX export a priority? My impression is the community has moved away from this

**Comment by adamjstewart** [comment]
I'm working on a funding proposal that requires ONNX export, but I think they only care about the model, not the preprocessing. PyTorch is 90% what people use for training, but other libraries are nice for lightweight deployment.

**Comment by robmarkcole** [comment]
Right, I think onnx could exist along an `extra` pathway with PyTorch being the default

**Comment by isaaccorley** [comment]
Actually I took a look at the refactor PR and this does seem to fix the previous export issues I raised. Will take another look

**Comment by edgarriba** [comment]
If there are other requests, let is know as we have some contributors willing to help

**Comment by calebrob6** [comment]
> Actually I took a look at the refactor PR and this does seem to fix the previous export issues I raised. Will take another look

Just took a look and onnx exports work great in kornia>0.9 (currently using RC1).

If you do `pip install --pre "kornia>=0.9.0rc1" onnx onnxruntime`

Then this all works:
```
import tempfile
import warnings
from pathlib import Path

import kornia
import kornia.augmentation as K
import numpy as np
import onnxruntime as ort
import torch
from torch import Tensor, nn

def export_to_onnx(pipeline: nn.Module, inputs: Tensor) -> bytes:
    """Export a preprocessing pipeline with the legacy ONNX exporter."""
    with tempfile.TemporaryDirectory() as directory:
        path = Path(directory) / 'preprocessing.onnx'
        with warnings.catch_warnings():
            warnings.simplefilter('ignore', DeprecationWarning)
            warnings.simplefilter('ignore', torch.jit.TracerWarning)
            warnings.filterwarnings(
                'ignore', message='To copy construct from a tensor.*'
            )
            torch.onnx.export(
                pipeline,
                (inputs,),
                path,
                opset_version=20,
                input_names=['input'],
                output_names=['output'],
                dynamo=False,
            )
        return path.read_bytes()


def main() -> None:
    """Export the pipeline and compare each result with eager execution."""
    torch.manual_seed(0)
    pipeline = K.AugmentationSequential(
        K.Normalize(
            mean=torch.tensor([0.1, 0.2, 0.3]), std=torch.tensor([0.9, 0.8, 0.7])
        ),
        K.Resize((40, 40)),
        K.CenterCrop((32, 32)),
        data_keys=['input'],
    ).eval()
    inputs = torch.rand(2, 3, 48, 48)

    with torch.inference_mode():
        eager_output = pipeline(inputs)

    exported = torch.export.export(pipeline, (inputs,))
    with torch.inference_mode():
        exported_output = exported.module()(inputs)
    torch.testing.assert_close(exported_output, eager_output)

    onnx_model = export_to_onnx(pipeline, inputs)
    session = ort.InferenceSession(onnx_model, providers=['CPUExecutionProvider'])
    onnx_output = session.run(['output'], {'input': inputs.numpy()})[0]
    np.testing.assert_allclose(onnx_output, eager_output.numpy(), rtol=1e-5, atol=1e-5)

    print(f'Kornia {kornia.__version__}')
    print(f'PyTorch {torch.__version__}')
    print(f'Output shape: {tuple(eager_output.shape)}')
    print('torch.export output matches eager output')
    print('ONNX Runtime output matches eager output')


if __name__ == '__main__':
    main()
```
